numpy/core/src/multiarray/multiarraymodule.c

```diff
@@ -1597,6 +1600,10 @@ _array_fromobject(PyObject *NPY_UNUSED(ignored), PyObject *args, PyObject *kws)
         return NULL;
     }
 
+    array_function_result = array_implement_c_array_function("array", args, kws);
+    if (array_function_result != NULL)
```

Comment:
How do you distinguish "not implemented" vs "threw an exception" here?

Reply:
Hmm, good point, I guess I'm not distinguishing? TBH, I'm not very experienced with writing C code for Python, I'll have to look at how to do it properly here.

Reply:
I'm really not very sure how the exception handling works in C, but without any modifications to the code, the exception from downstream library is passed through:

```python
In [1]: import numpy, cupy

In [2]: a = cupy.zeros(1)

In [3]: b = numpy.array(2, order='X', like=a)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-3-4c3c8e3f676e> in <module>
----> 1 b = numpy.array(2, order='X', like=a)

cupy/core/core.pyx in cupy.core.core.ndarray.__array_function__()

/datasets/pentschev/miniconda3/envs/rn-102-0.15.0b200717/lib/python3.7/site-packages/cupy/creation/from_data.py in array(obj, dtype, copy, order, subok, ndmin)
     41
     42     """
---> 43     return core.array(obj, dtype, copy, order, subok, ndmin)
     44
     45

cupy/core/core.pyx in cupy.core.core.array()

cupy/core/core.pyx in cupy.core.core.array()

cupy/core/core.pyx in cupy.core.core._send_object_to_gpu()

ValueError: order must be one of 'C', 'F', 'A', or 'K' (got 'X')
```

Is that mechanism "hidden" somewhere? Do I still miss anything w.r.t. proper exception handling in the case above?

Reply:
Python's C API (and NumPy internally) often uses `NULL` or `-1` as a sentinel value to indicate that an exception was raised, but the exception itself is kept track of via global state:
https://docs.python.org/3/c-api/exceptions.html

At least in principle, it is possible to raise exception from any code that uses Python's C API. So you almost always need to explicitly keep track of errors. (C Python is kind of like Go in this way)

Reply:
All that makes sense @shoyer , thanks for the details on exception handling with the C API! What I don't really understand is why the exception from my previous post got raised, even though I'm not currently handling it. Any ideas on where/how that ended up being handled?

Reply:
I've implemented exception handling with the error-related objects being passed as double pointers, see https://github.com/numpy/numpy/pull/16935/commits/58bbd08f922862ffad054ec920524ba7cfe1e7c6 . It seems to be done correctly, as I was able to get some `NotImplemented` errors with CuPy that I couldn't before, specifically by passing `subok=True` as in `np.array([1,2,3], subok=True, like=cupy.array(1))`, because `subok` isn't implemented by CuPy.

Reply:
Sorry @seberg , I replied to this comment in the main thread because GH weirdly shown it as a regular comment for me in the "Conversation" tab, but I see it in the correct place in the "Files changed" tab now.

The `NotImplemented` error I mentioned was just one of the exceptions I wasn't getting and now am able to get, but it would apply to any other downstream library exceptions, so I think the current solution seems to be more versatile.

Reply:
Oh, right, there is the issue that CuPy may (probably incorrectly) return `NotImplemented`, so we would have to use our own singleton.  Lets defer it...  I think there may be some merit in such redesign somewhere, possibly even for the typical python side wrappers, but right now I am unclear on what it would be. 

One more Nitpick though: The current `array_implement_c_array_function` actually hard-codes the `like` kwarg and only works for array-creation functions, maybe we should rename it and adjust the comment to that? (I guess a first version may have been different)

Reply:
Done in https://github.com/numpy/numpy/pull/16935/commits/9e0a7b5413bcac841e06fce84545fae0a4ece05d .